aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,607 @@
+#pragma omp parallel for private(k)
+    for (k = 0; k < numChannels; k++) {
+      int64_t finalOutputOffset = nOutputOffset + k * owidth * oheight;
+      int64_t finalInputOffset = nInputOffset + k * inputWidth * inputHeight;
+      scalar_t* output_p_k = rawOutput + finalOutputOffset;
+      scalar_t* input_p_k = rawInput + finalInputOffset;
+      int64_t* ind_p_k = rawIndices + finalInputOffset;
+
+      int64_t maxp;
+      for (int64_t i = 0; i < inputHeight; i++) {
+        for (int64_t j = 0; j < inputWidth; j++) {
+          maxp = ind_p_k[i * inputWidth + j];
+          if (maxp < 0 || maxp >= owidth * oheight) {
+#pragma omp critical
+            {
+              has_error = true;
+              error_index = maxp;
+            }
+          } else {
+            output_p_k[maxp] = input_p_k[i * inputWidth + j];
+          }
+        }
+      }
+    }
+    if (has_error) {
+      AT_ERROR(
+          "Found an invalid max index: ",
+          error_index,
+          " (output volumes are of size ",
+          oheight,
+          "x",
+          owidth);
+    }
+  }
+  return output;
+}
+
+Tensor& max_unpooling2d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& self_,
+    const Tensor& indices_,
+    IntList output_size) {
+  auto oheight = output_size[0];
+  auto owidth = output_size[1];
+  AT_CHECK(
+      indices_.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type int64");
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+  AT_CHECK(
+      (self_.ndimension() == 3 || self_.ndimension() == 4),
+      "Input to max_unpooling2d should be a 3d or 4d Tensor");
+  AT_CHECK(
+      self_.sizes() == indices_.sizes(),
+      "Shape of indices should match shape of input");
+
+  AT_CHECK(self_.numel() > 0, "Input must be non-empty");
+
+  auto self = self_.contiguous();
+  auto indices = indices_.contiguous();
+
+  if (self.ndimension() == 3) {
+    int64_t numBatch = 1;
+    int64_t numChannels = self.size(0);
+    output.resize_({numChannels, oheight, owidth});
+  } else {
+    int64_t numBatch = self.size(0);
+    int64_t numChannels = self.size(1);
+    output.resize_({numBatch, numChannels, oheight, owidth});
+  }
+  output.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES(
+      self.type(), "max_unpooling2d_forward_out_cpu_frame", ([&] {
+        max_unpooling2d_forward_out_cpu_frame<scalar_t>(
+            output, self, indices, owidth, oheight);
+      }));
+  return output;
+};
+
+Tensor max_unpooling2d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  auto output = at::empty({0}, self.options());
+  max_unpooling2d_forward_out_cpu(output, self, indices, output_size);
+  return output;
+}
+
+template <typename scalar_t>
+Tensor max_unpooling3d_forward_out_cpu_frame(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& indices,
+    int64_t oT,
+    int64_t oW,
```

Comment:
This looks like the wrong order to me. You're invoking this as:

```
            stride[0],
            stride[1],
            stride[2],
```

no?

Also below

Reply:
Actually, let me be more precise: the `o` arguments are handled correctly because you also reordered them at the call site, but the `d` and `p` arguments are wrong because the call site is not reordered. Honestly, it will be better and less error prone if you always put all arguments in the correct order.